### PR TITLE
[DUOS-1950][risk=no] Added study information component to data submission form page

### PR DESCRIPTION
**Pull request by lu-c**
----
Have you read [Terra's Contributing Guide](https://github.com/DataBiosphere/terra-ui/wiki/Contributor-Guide) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/components/SearchSelectOrText.js
Comment (lu-c) [inline src/components/SearchSelectOrText.js:25]: Not sure if this changes the searchSelectOrText behavior too much, it seemed okay to me, but could use confirmation!

File: cypress/component/DataSubmissionStudyInformation/ds_study_information.spec.js
Comment (rushtong) [inline cypress/component/DataSubmissionStudyInformation/ds_study_information.spec.js:5]: Codacy has a fix for the issue: [Strings must use singlequote.](https://app.codacy.com/gh/DataBiosphere/duos-ui/pullRequest?prid=9974908)
```suggestion
import DataSubmissionStudyInformation from '../../../src/components/data_submission/ds_study_information';
```

File: cypress/component/DataSubmissionStudyInformation/ds_study_information.spec.js
Comment (rushtong) [inline cypress/component/DataSubmissionStudyInformation/ds_study_information.spec.js]: Codacy has a fix for the issue: [Strings must use singlequote.](https://app.codacy.com/gh/DataBiosphere/duos-ui/pullRequest?prid=9974908)
```suggestion
import DataSubmissionStudyInformation from '../../../src/components/data_submission/ds_study_information';
```

File: src/components/data_submission/ds_study_information.js
Comment (connorlbark) [inline src/components/data_submission/ds_study_information.js:9]: should this file be snake case too? maybe this file should be `DataSubmissionStudyInformation`

File: src/pages/DataSubmissionForm.js
Comment (connorlbark) [inline src/pages/DataSubmissionForm.js]: use `setFormData`?

File: src/components/data_submission/ds_common.js
Comment (connorlbark) [inline src/components/data_submission/ds_common.js]: I really like the idea of making all of these generic components! Should we move this to somewhere where they can be used more broadly? Like just in the `Components` folder with a more descriptive name? Maybe in a separate ticket we could try to normalize the creation of preexisting form fields (in e.g. the researcher profile), as I find that consistent styling has been an issue - this would help mitigate that in forms.

File: src/components/data_submission/ds_common.js
Comment (connorlbark) [inline src/components/data_submission/ds_common.js]: what is the add new button for? maybe this should exist outside of `FormTable` since the component could be used for more things.

File: src/components/data_submission/ds_common.js
Comment (connorlbark) [inline src/components/data_submission/ds_common.js]: maybe we could make these strings constants and export them so that you don't ever have to use a raw string to specify things

**Review by connorlbark — COMMENTED** [review]
haven't ran it yet - but it looks great! biggest thing is i think it would be cool to make the FormField stuff more generic and able to used in more contexts.

File: src/pages/DataSubmissionForm.js
Comment (lu-c) [reply src/pages/DataSubmissionForm.js]: I think it ultimately should, but the ticket said to just make it an onChange method that will have a separate ticket to set

File: src/components/data_submission/ds_common.js
Comment (lu-c) [reply src/components/data_submission/ds_common.js]: The "Add new" button adds a new row to the formTable, so its functionality is tied pretty strictly to that!

File: src/components/data_submission/ds_common.js
Comment (lu-c) [reply src/components/data_submission/ds_common.js]: Good idea, i'll add that!

File: src/components/data_submission/dsStudyInformation.js
Comment (lu-c) [inline src/components/data_submission/dsStudyInformation.js]: Should I be using `dacUserId` or `userId`?

File: src/components/data_submission/ds_study_information.js
Comment (lu-c) [reply src/components/data_submission/ds_study_information.js:9]: Updated!

File: src/components/SearchSelectOrText.js
Comment (lu-c) [inline src/components/SearchSelectOrText.js:25]: Not sure if this changes the searchSelectOrText behavior too much, it seemed okay to me, but could use confirmation!

File: src/components/data_submission/ds_common.js
Comment (lu-c) [reply src/components/data_submission/ds_common.js]: I have so many thoughts about this! We should try to do as much as possible through css for normalizing the appearances of form fields, but where there is a necessary html component (like checkboxes and sliders), i would definitely want to create generic form components for it!

I think the components folder is really hard to browse for existing items though, and could use a little restructuring? Like, generic reuseable components should be grouped, and specific (potentially single purpose use) components shoudl also be grouped separately.

Maybe we could make a common/ colder inside components? Or it could exist outside of components as it's own `common/` too?

File: cypress/component/DataSubmissionStudyInformation/ds_study_information.spec.js
Comment (connorlbark) [inline cypress/component/DataSubmissionStudyInformation/ds_study_information.spec.js]: maybe the overall folder could be called `DataSubmission` so all the component tests for the overall functionality are in one place ?

Comment (lu-c) [reply cypress/component/DataSubmissionStudyInformation/ds_study_information.spec.js]: I'm sorry, I don't understand your suggestion - What do you mean?

File: src/components/data_submission/ds_study_information.js
Comment (JVThomas) [inline src/components/data_submission/ds_study_information.js]: Ignoring the exhaustive-deps rule is frowned upon. Here's a thread on Github with gaeron (Dan Abramov, software engineer on the React team) explaining his concerns
https://github.com/facebook/create-react-app/issues/6880

Basically turning the rule off causes developers to miss out on code smells or logical issues with the method in the useEffect itself. In this case, we see that assigning the required dependencies results in an infinite re-render loop.

So what's causing the issue? The two dependencies on this useEffect are `onChange` and `updateUserAndFields`. `updateUserAndFields` is defined in this component, so the only way that can trigger this function to re-fire is if the props are updated. And the only thing on props right now is `onChange`. So somehow, onChange is being updated when this useEffect is fired, causing the component to re-render and fire the useEffect again, resulting in the infinite loop.

Okay, so let's look at the parent component `DataSubmissionForm` and see what's up there. onChange looks to be an anonymous function that is defined directly on the component as a prop. This means that if `DataSubmissionForm` re-renders, the `onChange` function is re-defined, meaning the props update, triggering the re-render.

That said, what's causing the parent component to re-render? Well the onChange is updating `formData` with a cloned value. A cloned value means a new object reference. React triggers re-renders if props or state variables update their values. If the variable is an object, then React will compare object references (not the contents) to see if they are the same value. 

In this instance a cloned copy is a new object which means a different reference. This will trigger a re-render on the parent and start the infinite loop you're seeing. To avoid the complete page re-render, update the target keys on `formData`.

I know React promotes immutability, but it does so with regards to `this.state` itself (meaning you shouldn't directly update state unless using a hook). Updating `formData` here is fine, especially considering that each sub-component is going to rely on formData as a source of truth for data and rendering triggers.

For further reference, here's another post explaining the above in greater detail
https://blog.logrocket.com/immutability-in-react-ebe55253a1cc/

Comment (lu-c) [reply src/components/data_submission/ds_study_information.js]: Thanks for the explanation! Updated the component to not have an empty deps array, and cleaned up the sideeffects fallout!

File: src/components/forms/forms.js
Comment (lu-c) [inline src/components/forms/forms.js:33]: created this function bc long term, it might be useful to add different types of common validators (like Array.minLength)

File: src/components/data_submission/ds_study_information.js
Comment (JVThomas) [inline src/components/data_submission/ds_study_information.js]: Nit: The props object for hyperscript functions can process an `isRendered` key for conditional rendering. so you can do something like...

```
h(FormField, {
      id: 'dataSubmitterEmail',
      title: 'Data Submitter Email',
      defaultValue: user?.email,
      disabled: true,
      isRendered: !isEmpty(user),
      onChange
    })
```

and have the same effect.

File: src/pages/DataSubmissionForm.js
Comment (JVThomas) [inline src/pages/DataSubmissionForm.js]: Nit: stray console.log

**Comment by JVThomas** [comment]
The current functionality allows users to remove all elements on the File Type list.

Screen Shot 2022-08-16 at 11 27 54 AM

----

Screen Shot 2022-08-16 at 11 36 24 AM


Since we require there to be at least one element on the list (with all inputs filled), you'll need to either show an error if the list is empty or prevent them from removing all elements. You can approach this in two ways...

1. Only show the close/delete icon if there's more than one item
2. Show a validation error if there are no elements present in the list (show the error box below the table).

Considering that the other form components allow for validators, I think it's reasonable for the FormTable to take in validation props that it can apply to itself (like min size of `enter size here`). The fields themselves should also have validators defined to check if the input is empty.

File: src/components/forms/forms.js
Comment (connorlbark) [inline src/components/forms/forms.js]: this string should be a prop so that the field form can be used generically

File: src/components/data_submission/dsStudyInformation.js
Comment (rushtong) [reply src/components/data_submission/dsStudyInformation.js]: `userId` ... dac user id is deprecated in all of the APIs.

File: src/components/forms/forms.js
Comment (lu-c) [reply src/components/forms/forms.js]: done!

File: src/pages/DataSubmissionForm.js
Comment (lu-c) [reply src/pages/DataSubmissionForm.js]: i actually kind of liked it there... just while there was no other confirmation that the code was working as intended. Would you be okay with keeping this until the feature is completed?

Comment (JVThomas) [reply src/pages/DataSubmissionForm.js]: Sure, I was worried Codacy was going to freak out about it, but apparently not

**Comment by lu-c** [comment]
Found a bug with deleting rows (value updates, but not ui), going to tackle that next!

File: src/components/forms/forms.js
Comment (connorlbark) [inline src/components/forms/forms.js]: `validators.includes(FormValidators.REQUIRED)`?

Comment (connorlbark) [reply src/components/forms/forms.js]: same comment other usages of this

File: src/pages/DataSubmissionForm.js
Comment (connorlbark) [inline src/pages/DataSubmissionForm.js:17]: I like the is valid approach!

**Review by connorlbark — APPROVED** [review]
Everything works super well locally! Couple nits but overall I would not mind getting this merged so the rest of us can use the FormField work (which is awesome: great work on that). My only concern is that I was confused on the MULTITEXT fields. I didn't realize I had to hit `Enter` for it to update / submit the value. Maybe there should be a small '+' or something on the righthand side? I would be ok with tackling this now or by doing it in a separate small ticket so that we can free up others to work on the form.

File: src/components/forms/forms.js
Comment (lu-c) [reply src/components/forms/forms.js]: updated!

**Comment by lu-c** [comment]
> Everything works super well locally! Couple nits but overall I would not mind getting this merged so the rest of us can use the FormField work (which is awesome: great work on that). My only concern is that I was confused on the MULTITEXT fields. I didn't realize I had to hit `Enter` for it to update / submit the value. Maybe there should be a small '+' or something on the righthand side? I would be ok with tackling this now or by doing it in a separate small ticket so that we can free up others to work on the form.

Yeah, sure, I can add a button for it! Still waiting on a second approval, but I won't let that hold this pr back. ;) Will merge whether or not the button exists, and figure out adding that button when it can

**Review by JVThomas — APPROVED** [review]
The code looks great! The form is validating the fields as you would expect. The form components themselves will be extremely useful for the rest of the form (and any other form in the future).

**Pull request merged**
